source/cppfs/source/windows/LocalFileWatcher.cpp

```diff
@@ -0,0 +1,196 @@
+
+#include <cppfs/windows/LocalFileWatcher.h>
+
+#include <algorithm>
+
+#include <Windows.h>
+
+#include <cppfs/windows/LocalFileSystem.h>
+
+
+namespace
+{
+    struct LocalWatcher
+    {
+        OVERLAPPED ovl;
+        std::shared_ptr<VOID> event;
+        DWORD buffer[16384];
+    };
+
+    // NB! Using a std::mutex resulted in spurious deadlocks using VS2017, whilst
+    // using CRITICAL_SECTION works fine.
+    struct ScopedCriticalSection
+    {
+        LPCRITICAL_SECTION m_pcs;
+        ScopedCriticalSection(LPCRITICAL_SECTION pcs) : m_pcs(pcs) {
+            ::EnterCriticalSection(m_pcs);
+        }
+        ~ScopedCriticalSection() {
+            ::LeaveCriticalSection(m_pcs);
+        }
+    };
+}
+
+
+namespace cppfs
+{
+
+
+LocalFileWatcher::LocalFileWatcher(FileWatcher * fileWatcher, std::shared_ptr<LocalFileSystem> fs)
+: AbstractFileWatcherBackend(fileWatcher)
+, m_fs(fs)
+, m_waitStopEvent(::CreateEvent(NULL, TRUE, FALSE, NULL), ::CloseHandle)
+, m_watchersCS(new CRITICAL_SECTION())
+{
+    ::InitializeCriticalSection((LPCRITICAL_SECTION)m_watchersCS.get());
+}
+
+LocalFileWatcher::~LocalFileWatcher()
+{
+    ::DeleteCriticalSection((LPCRITICAL_SECTION)m_watchersCS.get());
+    m_watchers.clear();
+}
+
+AbstractFileSystem * LocalFileWatcher::fs() const
+{
+    return static_cast<AbstractFileSystem *>(m_fs.get());
+}
+
+void LocalFileWatcher::add(FileHandle & fh, unsigned int events, RecursiveMode recursive)
+{
+    HANDLE hDir = ::CreateFileA(fh.path().c_str(),       // pointer to the directory name
+                                      FILE_LIST_DIRECTORY,  // access (read/write) mode
+                                      FILE_SHARE_READ       // share mode
+                                      | FILE_SHARE_WRITE | FILE_SHARE_DELETE,
+                                      NULL,                       // security descriptor
+                                      OPEN_EXISTING,              // how to create
+                                      FILE_FLAG_BACKUP_SEMANTICS  // file attributes
+                                      | FILE_FLAG_OVERLAPPED,
+                                      NULL);  // file with attributes to copy
+    if (hDir == INVALID_HANDLE_VALUE)
+        throw std::runtime_error("failed to create directory listener");
+
+    std::shared_ptr<void> hDirectory(hDir, ::CloseHandle);
+    std::shared_ptr<void> hEvent(::CreateEvent(NULL, TRUE, FALSE, NULL), ::CloseHandle);
+    if (!hEvent)
+        throw std::runtime_error("failed creating wait event");
+
+    auto lw = std::make_shared<LocalWatcher>();
```

Comment:
Are the shared pointers here ever shared between several objects, or are they unique? (seems to me that way). Maybe we can use unique_ptr, or even just plain handles?

Reply:
shared_ptr is used because you can specify the delete functions for the object, which in case of Windows HANDLE is the ::CloseHandle function. Out of pure convenience, and to avoid dangling handles.